client-libraries/rust/src/plugin_utils/ingest.rs

```diff
@@ -0,0 +1,1065 @@
+//! }
+//!
+//! #[tokio::main]
+//! async fn main() -> Box<dyn std::error::Error> {
+//!   init_tracing!();
+//!   let cfg = Config::<MyConfig>::load("MY_PLUGIN_")?;
+//!   let mut client = cfg.connect_and_authenticate().await?;
+//!   
+//!   let timeline_id = TimelineId::allocate();
+//!   client.switch_timeline(timeline_id).await?;
+//!   client.send_timeline_attrs("tl", [
+//!     ("attr1", 42.into())
+//!     ("attr2", "hello".into())
+//!   ]).await?;
+//!
+//!   client.send_event("ev", [
+//!     ("attr1", 42.into())
+//!     ("attr2", "hello".into())
+//!   ]).await?;
+//! }
+//! ```
+
+use crate::{
+    api::{AttrVal, Nanoseconds, TimelineId},
+    auth_token::AuthToken,
+    ingest_client::{dynamic::DynamicIngestClient, IngestClient, ReadyState},
+    ingest_protocol::InternedAttrKey,
+    reflector_config::{
+        AttrKeyEqValuePair, ConfigLoadError, SemanticErrorExplanation, TomlValue, TopLevelIngest,
+        CONFIG_ENV_VAR,
+    },
+};
+use serde::{de::DeserializeOwned, Deserialize, Serialize};
+use std::{
+    collections::BTreeMap,
+    env,
+    path::{Path, PathBuf},
+    str::FromStr as _,
+    time::{Duration, SystemTime},
+};
+use url::Url;
+
+/// Plugin configuration structure; contains both common elements, and
+/// plugin-specific elements, based on the type param `T`.
+pub struct Config<T> {
+    /// Common ingest configuration; mostly connection-related.
+    pub ingest: TopLevelIngest,
+
+    /// The plugin-specific portion of the configuration.
+    pub plugin: T,
+
+    /// The client connection timeout. This is automatically used when
+    /// you call `connected_and_authenticate`.
+    pub client_timeout: Option<Duration>,
+
+    /// `timeline.run_id` will be set to this value for all created
+    /// timelines.
+    pub run_id: String,
+
+    /// If Some(...), `timeline.time_domain` will be set to this value
+    /// for all created timelines.
+    pub time_domain: Option<String>,
+}
+
+#[derive(Deserialize)]
+struct EnvConfig {
+    // MODALITY_CLIENT_TIMEOUT Environment variable
+    modality_client_timeout: Option<f32>,
+
+    // MODALITY_RUN_ID Environment variable
+    modality_run_id: Option<String>,
+
+    // MODALITY_TIME_DOMAIN Environment variable
+    modality_time_domain: Option<String>,
+}
+
+impl<T: Serialize + DeserializeOwned> Config<T> {
+    /// Load configuration from config file given in
+    /// `MODALITY_REFLECTOR_CONFIG` as well as from other environment
+    /// variables (see module documentation). The returned `Config`
+    /// structure represents the fully reconcicled configuration.
+    ///
+    /// * `env_prefix`: The prefix used for environment variable based
+    ///   settings for members of the configuration struct (type
+    ///   param `T`).
+    pub fn load(env_prefix: &str) -> Result<Config<T>, Box<dyn std::error::Error>> {
+        Self::load_custom(env_prefix, |_, _| None)
+    }
+
+    /// Load configuration, like `Config::load`, but allows passing a
```

Comment:
Should bracket `[]` any references to types/etc defined in this crate so rustdoc will produce links.